addons/l10n_fr_hr_holidays/models/hr_leave.py

```diff
@@ -0,0 +1,150 @@
+        company_working_days = self._get_working_hours(company_calendar)
+        if not self._calendar_works_on_date(CalendarAttendance, company_calendar, company_working_days, date):
+            return False
+        employee_working_days = self._get_working_hours(employee_calendar)
+        if self._calendar_works_on_date(CalendarAttendance, employee_calendar, employee_working_days, date):
+            return False
+        return True
+
+    def _get_fr_new_date_to(self, date_to, employee_calendar, company_calendar):
+        employee_working_days = self._get_working_hours(employee_calendar)
+        company_working_days = self._get_working_hours(company_calendar)
+
+        CalendarAttendance = self.env['resource.calendar.attendance']
+        date_target = date_to
+        if isinstance(date_to, str):
+            date_target = datetime.fromisoformat(date_to)
+        new_date_to = date_target
+        date_target += relativedelta(days=1)
+        while not self._calendar_works_on_date(CalendarAttendance, employee_calendar, employee_working_days, date_target):
+            if self._calendar_works_on_date(CalendarAttendance, company_calendar, company_working_days, date_target):
+                new_date_to = date_target
+            date_target += relativedelta(days=1)
+
+        return new_date_to
+
+    def _get_fr_number_of_days(self, employee, date_from, date_to, employee_calendar, company_calendar):
+        self.ensure_one()
+
+        self.l10n_fr_date_to = False
+        # We can fill the holes using the company calendar as default
+        # What we need to compute is how much we will need to push date_to in order to account for the lost days
+        # This gets even more complicated in two_weeks_calendars
+        employee_working_days = self._get_working_hours(employee_calendar)
+
+        CalendarAttendance = self.env['resource.calendar.attendance']
+        if self.request_unit_half:
+            # In normal workflows request_unit_half implies that date_from and date_to are the same
+            # request_unit_half allows us to choose between `am` and `pm`
+            # In a case where we work from mon-wed and request a half day in the morning
+            # we do not want to push date_to since the next work attendance is actually in the afternoon
+            date_from_weektype = str(CalendarAttendance.get_week_type(date_from))
+            date_from_dayofweek = str(date_from.weekday())
+            # Fetch the attendances we care about
+            attendance_ids = employee_calendar.attendance_ids.filtered(lambda a:
+                a.dayofweek == date_from_dayofweek
+                and (not employee_calendar.two_weeks_calendar or a.week_type == date_from_weektype))
+            if len(attendance_ids) == 2 and self.request_date_from_period == 'am':
+                # The employee took the morning off on a day where he works the afternoon aswell
```

Comment:
You can only do the logic for morning leaves as the rest of the computation will not be used in the other case:

```suggestion
            # we do not want to push date_to since the next work attendance is actually in the afternoon
            if self.request_date_from_period == 'am'
                date_from_weektype = str(CalendarAttendance.get_week_type(date_from))
                date_from_dayofweek = str(date_from.weekday())
                # Fetch the attendances we care about
                attendance_ids = employee_calendar.attendance_ids.filtered(lambda a:
                    a.dayofweek == date_from_dayofweek
                    and (not employee_calendar.two_weeks_calendar or a.week_type == date_from_weektype))
                if len(attendance_ids) == 2:
                # The employee took the morning off on a day where he works the afternoon aswell
```